notes/force-move-implementation.md

```diff
@@ -0,0 +1,178 @@
+- We want it to be efficient to calculate without having to pass in the full `storedState`
+
+### Proposal for `ChannelStorageHash`
+
+The `ChannelStorageHash` is the hash of:
+
+- `turnNumRecord`
+- `finalizesAt`
+- `stateHash`
+- `challengerAddress`
+- `outcomeHash`
+
+**Why include the `outcomeHash`?** Although the `outcome` is included in the `state`, we include the `outcomeHash` at the top level of the `channelStorageHash` to make it easier for the PushOutcome method to prove what the outcome of the channel was. The tradeoff here is that the methods need to make sure they have the data to calculate it - which adds at most a `bytes32` to their calldata.
+
+### What else do the methods need?
+
+- ForceMove
+  - Needs to be passed a sequence of up to n full states
+  - The states all share the same `chainId`, `participants`, `channelNonce` and `appDefinition`
+  - The `turnNum` must increase by 1 each time
+  - `isFinal` can transition from 0 to 1 at one point
+  - Need `outcome` and `appData` for each state
+  - Needs n distinct signatures
+- Respond
+  - Needs one full state
+- RespondFromAlternative
+  - Needs to be passed a sequence of n full states
+- Refute
+  - Needs a proof of signature of higher state
+  - So we just need to see the turnNum
+- Conclude
+  - Can deduce all states after the first (as they're the same apart from an increasing turn number)
+  - Needs to know is final
+  - Doesn't need the details of outcome or appData
+  - Needs to set the outcomehash
+
+From this we can say the following about the way we chose to hash states:
+
+- It should be efficient to calculate multiple state hashes for consecutive states, for signature verification => we want to hash all of the fixed parts together (actually if there are < 3 states, it's probably cheaper to add the app def at the top level instead of another hash)
+- We're always going to have to calculate the channelId anyway => the hash of the fixed parts might as well contain the hash of the channelId
+- We want to be able to prove that the outcome is part of the channelStatus
+- Want to be able to succinctly prove that someone signed a later state => turnNum should be separate
+
+### Proposal for state hash
+
+The hash of the state that the participants sign should be the hash of the following:
+
+- TurnNum
+- isFinal
+- ChannelId
+  - ChainId
+  - Participants
+  - ChannelNonce
+- VariablePartHash
+  - AppDefinition
+  - OutcomeHash
+  - AppData
+
+**Why include the ChannelId separately?** We have to calculate the `channelId` for every single operation anyway. Given that we have it it's cheaper to hash in the hash, rather than the individual components again.
+
+**Why not include the `AppDefinition` in the `ChannelId`?** The `AppDefinition` is fixed as part of the transition rules, so it seems like it would make sense to include it in the `ChannelId`. This would rule out participants being able to collaboratively upgrade the app without refunding the channel through.
```

Comment:
It could still be possible to collaboratively upgrade the app via an on-chain transaction if using something like ZeppelinOS.

Reply:
Good point. Do you think this means that the `AppDefinition` belongs in the `ChannelId`? It seems to me that there's still no advantage to putting it in, when it doesn't need to be there.

Reply:
I think it's better to not include it and ensure the client uses a unique channel nonce per